raiden/storage/sqlite.py

```diff
@@ -1,33 +1,58 @@
 import sqlite3
-import threading
 from contextlib import contextmanager
+from dataclasses import dataclass
 from datetime import datetime
 from enum import Enum
 from pathlib import Path
 
+from typing_extensions import Literal
+
 from raiden.constants import RAIDEN_DB_VERSION, SQLITE_MIN_REQUIRED_VERSION
 from raiden.exceptions import InvalidDBData, InvalidNumberInput
 from raiden.storage.serialization import SerializationBase
+from raiden.storage.ulid import ULID, ULIDMonotonicFactory
 from raiden.storage.utils import DB_SCRIPT_CREATE_TABLES, TimestampedEvent
 from raiden.transfer.architecture import Event, State, StateChange
 from raiden.utils import get_system_spec
 from raiden.utils.typing import (
     Any,
     Dict,
-    EventID,
+    Generic,
     Iterator,
     List,
     NamedTuple,
+    NewType,
     Optional,
     RaidenDBVersion,
-    SnapshotID,
-    StateChangeID,
-    T_StateChangeID,
     Tuple,
+    Type,
+    TypeVar,
     Union,
-    typecheck,
 )
 
+StateChangeID = NewType("StateChangeID", ULID)
+SnapshotID = NewType("SnapshotID", ULID)
+EventID = NewType("EventID", ULID)
+ID = TypeVar("ID", StateChangeID, SnapshotID, EventID)
+
+
+@dataclass
+class Range(Generic[ID]):
+    """Inclusive range used to filter database entries."""
+
+    first: ID
+    last: ID
+
+    def __post_init__(self) -> None:
+        # Because the range is inclusive, the values can be equal
+        if self.first >= self.last:
+            raise ValueError("last must be larger or equal to first")
+
+
+FIRST_STATECHANGE_ULID = StateChangeID(ULID((0).to_bytes(16, "big")))
+LAST_STATECHANGE_ULID = StateChangeID(ULID((2 ** 128 - 1).to_bytes(16, "big")))
```

Comment:
These are the first and last possible values for the IDs, not the first and last value in the database.

Reply:
```suggestion
LOW_STATECHANGE_ULID = StateChangeID(ULID((0).to_bytes(16, "big")))
HIGH_STATECHANGE_ULID = StateChangeID(ULID((2 ** 128 - 1).to_bytes(16, "big")))
```